url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/logic/LogikusParser.java

```diff
@@ -0,0 +1,523 @@
+ * <code>Logikus</code> parser in practice. It uses 
+ * <code>LogikusFacade</code> to create a <code>Query</code>, 
+ * proves the query, and displays the query's variables for 
+ * each proof. As in Prolog, the Logikus development 
+ * environment prints "no" when no further proofs remain.
+ *
+ * @author Steven J. Metsker
+ *
+ * @version 1.0 
+ */
+public class LogikusParser {
+	protected Sequence structure;
+	protected Sequence expression;
+	protected Sequence list;
+
+	/*
+	 * Return a parser that recognizes the grammar:
+	 * 
+	 *    arg = expression | functor;
+	 */
+	protected Parser arg() {
+		Alternation a = new Alternation();
+		a.add(expression());
+		a.add(functor().setAssembler(new AtomAssembler()));
+		return a;
+	}
+
+	/**
+	 * Return a parser that recognizes the grammar:
+	 *
+	 * <blockquote><pre>
+	 *    axiom = structure (ruleDef | Empty);
+	 * </pre></blockquote>
+	 *
+	 * @return a parser that recognizes an axiom
+	 */
+	public Parser axiom() {
+		Sequence s = new Sequence("axiom");
+
+		s.add(structure());
+		Alternation a = new Alternation();
+		a.add(ruleDef());
+		a.add(new Empty());
+		s.add(a);
+
+		s.setAssembler(new AxiomAssembler());
+		return s;
+	}
+
+	/*
+	 * Using the given parser, this method composes a new
+	 * parser with the grammar:
+	 * 
+	 *     commaList(p) = p (',' p)*;
+	 *
+	 * The Logikus language uses this construction several 
+	 * times.
+	 */
+	protected static Sequence commaList(Parser p) {
+		Sequence commaP = new Track();
+		commaP.add(new Symbol(',').discard());
+		commaP.add(p);
+
+		Sequence s = new Sequence();
+		s.add(p);
+		s.add(new Repetition(commaP));
+		return s;
+	}
+
+	/**
+	 * Return a parser that recognizes the grammar:
+	 *
+	 * <blockquote><pre>
+	 *    comparison = operator '(' arg ',' arg ')';
+	 * </pre></blockquote>
+	 *
+	 * @return a parser that recognizes a comparison
+	 */
+	public Sequence comparison() {
+		Track t = new Track("comparison");
+		t.add(operator());
+		t.add(new Symbol('(').discard());
+		t.add(arg());
+		t.add(new Symbol(',').discard());
+		t.add(arg());
+		t.add(new Symbol(')').discard());
+		t.setAssembler(new ComparisonAssembler());
+		return t;
+	}
+
+	/**
+	 * Return a parser that recognizes the grammar:
+	 *
+	 * <blockquote><pre>
+	 *    condition = structure | not | evaluation | comparison | 
+	 *                list;
+	 * </pre></blockquote>
+	 *
+	 * @return a parser that recognizes a condition
+	 */
+	public Alternation condition() {
+		Alternation a = new Alternation("condition");
+		a.add(structure());
+		a.add(not());
+		a.add(evaluation());
+		a.add(comparison());
+		a.add(list());
+		return a;
+	}
+
+	/*
+	 * Return a parser that recognizes the grammar:
+	 * 
+	 *    divideFactor = '/' factor;
+	 */
+	protected Parser divideFactor() {
+		Sequence s = new Sequence("divideFactor");
+		s.add(new Symbol('/').discard());
+		s.add(factor());
+		s.setAssembler(new ArithmeticAssembler('/'));
+		return s;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 186:190, 333:337, 400:404, 496:500.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.